File: PDOO/p4/civitas_rb/lib/sorpresa_porjugador.rb
module Civitas
  
class Sorpresa_porjugador < Sorpresa
  
  def initialize(valor, texto)
    super(texto)
    @valor=valor
  end
  
  
  def aplicar_a_jugador(actual, todos)
    if (jugador_correcto(actual, todos))
      informe(actual, todos)
      s1 = Sorpresa_pagarcobrar.new(@valor * -1, "PAGARCOBRAR")
      for i in 0...actual
        s1.aplicar_a_jugador(i, todos)
      end
      for j in actual+1...todos.length
        s1.aplicar_a_jugador(j, todos)
      end
      s2 = Sorpresa_pagarcobrar.new(@valor * (todos.length - 1), "PAGARCOBRAR0")
      s2.aplicar_a_jugador(actual, todos)
    end
  end
  
  def to_string()
    return super
  end
  
end
end

File: PDOO/p4/Civitas/src/civitas/SorpresaPorJugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author aida
 */
public class SorpresaPorJugador extends Sorpresa {
    
    private int valor;
    
    @Override
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(actual, todos)){
            informe(actual, todos);
            SorpresaPagarCobrar s1 = new SorpresaPagarCobrar(valor * -1, "PAGARCOBRAR");
            for (int i = 0; i < actual; i++){
                s1.aplicarAJugador(i, todos);
            }
            for (int i = actual + 1; i < todos.size(); i++){
                s1.aplicarAJugador(i, todos);
            }
            SorpresaPagarCobrar s2 = new SorpresaPagarCobrar(valor * (todos.size() - 1), "PAGARCOBRAR");
            s2.aplicarAJugador(actual, todos);
        }
    }
    
    SorpresaPorJugador(int valor, String texto){
        //Para el resto de sorpresas
        super(texto);
        this.valor = valor;
    }
    
    @Override
    public String toString(){
        return "Sorpresa{" + "texto=" + super.getTexto() + ", valor=" + Integer.toString(valor) + "}";
    }
}
